src/coreclr/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -1252,22 +1252,31 @@ void CodeGen::genBaseIntrinsic(GenTreeHWIntrinsic* node)
         }
 
         case NI_Vector128_get_AllBitsSet:
-        case NI_Vector256_get_AllBitsSet:
-        {
             assert(op1 == nullptr);
-            if (varTypeIsFloating(baseType) && compiler->compOpportunisticallyDependsOn(InstructionSet_AVX))
+            if (compiler->compOpportunisticallyDependsOn(InstructionSet_AVX))
             {
-                // The immediate 8 means Equal (unordered, non-signaling)
-                // This is not available without VEX prefix.
-                emit->emitIns_SIMD_R_R_R_I(ins, attr, targetReg, targetReg, targetReg, 8);
+                emit->emitIns_SIMD_R_R_R(INS_pcmpeqd, attr, targetReg, targetReg, targetReg);
             }
             else
             {
-                assert(varTypeIsIntegral(baseType) || !compiler->compIsaSupportedDebugOnly(InstructionSet_AVX));
+                emit->emitIns_R_R(INS_pcmpeqd, attr, targetReg, targetReg);
+            }
+            break;
+
+        case NI_Vector256_get_AllBitsSet:
+            assert(op1 == nullptr);
+            if (compiler->compOpportunisticallyDependsOn(InstructionSet_AVX2))
+            {
                 emit->emitIns_SIMD_R_R_R(INS_pcmpeqd, attr, targetReg, targetReg, targetReg);
             }
+            else
+            {
+                assert(compiler->compIsaSupportedDebugOnly(InstructionSet_AVX));
+                emit->emitIns_SIMD_R_R_R(INS_xorps, EA_16BYTE, targetReg, targetReg, targetReg);
```

Comment:
We shouldn't need to zero the target register when using `8` as the immediate.

`8` is `VCMPEQ_UQSS` which is "compare equal, unordered, non-signalling" which effectively means that if `left` and `right` are the same or either is NaN, return `true` and do not throw for signalling NaN if floating-point exceptions are enabled".

Since `left` and `right` are both `target`, it means it will always return `true` (the same is technically true for `15`).

You can confirm this with a program such as:
```csharp
using System;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;

class Program
{
    static void Main(string[] args)
    {
        var nan = Vector128.Create(float.NaN);
        var one = Vector128.Create(1.0f);
        var inf = Vector128.Create(float.PositiveInfinity);

        Test(nan, FloatComparisonMode.UnorderedEqualNonSignaling);
        Test(nan, FloatComparisonMode.UnorderedTrueNonSignaling);

        Test(one, FloatComparisonMode.UnorderedEqualNonSignaling);
        Test(one, FloatComparisonMode.UnorderedTrueNonSignaling);

        Test(inf, FloatComparisonMode.UnorderedEqualNonSignaling);
        Test(inf, FloatComparisonMode.UnorderedTrueNonSignaling);
    }

    static void Test(Vector128<float> value, FloatComparisonMode mode)
    {
        Console.WriteLine($"{value,-41}, {mode,-28}: {Avx.Compare(value, value, mode)}");
    }
}
```

which prints:
```
<NaN, NaN, NaN, NaN>                     , UnorderedEqualNonSignaling  : <NaN, NaN, NaN, NaN>
<NaN, NaN, NaN, NaN>                     , UnorderedTrueNonSignaling   : <NaN, NaN, NaN, NaN>
<1, 1, 1, 1>                             , UnorderedEqualNonSignaling  : <NaN, NaN, NaN, NaN>
<1, 1, 1, 1>                             , UnorderedTrueNonSignaling   : <NaN, NaN, NaN, NaN>
<Infinity, Infinity, Infinity, Infinity> , UnorderedEqualNonSignaling  : <NaN, NaN, NaN, NaN>
<Infinity, Infinity, Infinity, Infinity> , UnorderedTrueNonSignaling   : <NaN, NaN, NaN, NaN>
```

Reply:
But why choosing `VCMPEQ_UQSS` over `VCMPTRUESS` when the latter does exactly (setting all the bits to ones) what we need?

Reply:
The idea of emitting `xorps` before `vcmptruess` was also explained in one of the links I mentioned earlier - it is there to break dependency on the prior value in the register. 

Also see https://www.agner.org/optimize/optimizing_assembly.pdf page 65

Reply:
> But why choosing VCMPEQ_UQSS over VCMPTRUESS when the latter does exactly (setting all the bits to ones) what we need?

For the scenario being considered; they both look to operate identically. I don't recall why `VCMPEQ_UQSS` was chosen originally over `VCMPTRUESS` but it was and is what we are emitting today.
I don't have a strong preference on which of these two are emitted provided they are always identical for `tgt, tgt, tgt`.

> The idea behind emitting xorps before vcmptruess was also explained in one of the links I mentioned earlier - it is there to break dependency on the prior value in the register.

We aren't doing this for any of the other comparisons or for intrinsics in general (outside LZCNT or POPCNT IIRC, because older Intel CPUs had a bug where it always treated it as having a dependency). Likewise, I don't think we should be unconditionally inserting a dependency breaking idiom here and instead this would be better handled by some general purpose check that identifies possible dependencies and therefore where inserting this would be beneficial.

Hardware intrinsics are used in performance oriented code paths and even if `xorps` is elided in the decoding step, it still takes up space in the decoder and is going to change the overall throughput of small-medium sized loops.
Additionally, Clang itself isn't emitting this idiom on haswell or newer (essentially any machine with AVX2+) and is not emitting it on skylake (which is what most of our benchmarks and even estimated cycle counts are based on).

Reply:
> Additionally, Clang itself isn't emitting this idiom on haswell or newer (essentially any machine with AVX2+) and is not emitting it on skylake (which is what most of our benchmarks and even estimated cycle counts are based on).

Right, for AVX2+ Clang emits `vpcmpeqd ymmReg, ymmReg, ymmReg` as I poined out in the first post and implemented for `NI_Vector256_get_AllBitsSet`.